### ISSUE TITLE
Failure to authenticate against GHE

### ISSUE THREAD

**Issue by jefflage** [octokit/octokit.js#1683]
Both of the following code snippets fail to authenticate agents our internal instance of GHE:
```javascript
async function getApp(app, url) {
    const jwt = app.getSignedJsonWebToken();

    const octokit = new Octokit({
        auth: jwt,
        baseUrl: url,
        log: console
    });

    return octokit.apps.getAuthenticated();
}

(async() => {
    const app = new App({
        id: 1,
        privateKey: privKey,
        baseUrl: url
    });

    const appInfo = await getApp(app, url);
    console.log("Got app: " + appInfo.data.name);
})();
```
```javascript
const octokit = new Octokit({
    authStrategy: createAppAuth,
    auth: { id, privateKey },
    baseUrl
});

(async() => {
    const appInfo = await octokit.apps.getAuthenticated();
    console.log(appInfo);
})();
```

With the following error: `UnhandledPromiseRejectionWarning: HttpError: Bad credentials`

A support case was opened with Enterprise Support (#628809)

However, if in the first code snipped, I dump the generated JWT to the console, and use that JWT in a basic `curl` command, the call authenticates successfully.  I'm thinking that perhaps Octokit/rest.js is not sending/setting the header.  I am basing this on the fact that no auth headers are dumped to the console but other headers are.

**Comment by Sid-ah** [comment]
In the first snippet you're not returning a promise hence why your await doesn't work on the iife (Immediately Invoked Function Expression). if you update the snippet to the following, it may fix the issue you're facing:
```js
function getApp(app, url) {
  const jwt = app.getSignedJsonWebToken();

  const octokit = new Octokit({
    auth: jwt,
    baseUrl: url,
    log: console,
  });

  return new Promise((resolve, reject) => {
    if (octokit.apps.getAuthenticated()) {
      resolve(octokit.apps.getAuthenticated());
    } else {
      reject("there was an error while trying to get authenticated");
    }
  });
}

(async () => {
  const app = new App({
    id: 1,
    privateKey: privKey,
    baseUrl: url,
  });

  const appInfo = await getApp(app, url);
  console.log("Got app: " + appInfo.data.name);
})();
 ```

I hope this help :D

**Comment by jefflage** [comment]
@Sid-ah According to the TypeSciprt typedefs bundled with Octokit/rest.js, that function does return a promise:

`getAuthenticated: {
            (params?: RequestParameters & EmptyParams): Promise<OctokitResponse<AppsGetAuthenticatedResponse>>;
            endpoint: EndpointInterface;
        };
`

**Comment by Sid-ah** [comment]
@jefflage you can also leave the first code snippet as it is and replace the following lines: 

```js
const authResult = await octokit.apps.getAuthenticated();
return authResult;
```

**Comment by gr2m** [comment]
Hi @jefflage, so far I was not able to reproduce the problem with the following code

```js
const { Octokit } = require("@octokit/rest");
const { createAppAuth } = require("@octokit/auth-app");

const octokit = new Octokit({
  authStrategy: createAppAuth,
  auth: {
    id: 123...,
    privateKey: `-----BEGIN RSA PRIVATE KEY-----
MIIEpAIBA...
...
...9NL/bP9Q==
-----END RSA PRIVATE KEY-----`,
  },
  baseUrl: "https://...",
});

(async () => {
  const appInfo = await octokit.apps.getAuthenticated();
  console.log(appInfo);
})();
```

Can you share the following

1. What GitHub Enterprise Server version are you using
2. Output of `npm ls | grep octokit`

**Comment by jefflage** [comment]
@gr2m Apologies for the delay:

1. GHE Version: 2.20.3
2.  Output from npm:
```
├─┬ @octokit/app@4.2.0
│ ├─┬ @octokit/request@5.3.4
│ │ ├─┬ @octokit/endpoint@6.0.0
│ │ │ ├── @octokit/types@2.5.1 deduped
│ │ ├── @octokit/request-error@2.0.0 deduped
│ │ ├── @octokit/types@2.5.1 deduped
├─┬ @octokit/auth-app@2.4.4
│ ├── @octokit/request@5.3.4 deduped
│ ├─┬ @octokit/request-error@2.0.0
│ │ ├── @octokit/types@2.5.1 deduped
│ ├─┬ @octokit/types@2.5.1
├─┬ @octokit/rest@17.1.4
│ ├─┬ @octokit/core@2.4.3
│ │ ├─┬ @octokit/auth-token@2.4.0
│ │ │ └── @octokit/types@2.5.1 deduped
│ │ ├─┬ @octokit/graphql@4.3.1
│ │ │ ├── @octokit/request@5.3.4 deduped
│ │ │ ├── @octokit/types@2.5.1 deduped
│ │ ├── @octokit/request@5.3.4 deduped
│ │ ├── @octokit/types@2.5.1 deduped
│ ├─┬ @octokit/plugin-paginate-rest@2.0.2
│ │ └── @octokit/types@2.5.1 deduped
│ ├── @octokit/plugin-request-log@1.0.0
│ └─┬ @octokit/plugin-rest-endpoint-methods@3.4.0
│   ├── @octokit/types@2.5.1 deduped
```

**Comment by gr2m** [comment]
I got access to a GHE (2.20.7) instance for testing.

I tried to reproduce your problem with the following code:

```js
const { readFileSync } = require("fs");

const { Octokit } = require("@octokit/core");
const { createAppAuth } = require("@octokit/auth-app");

const octokit = new Octokit({
  authStrategy: createAppAuth,
  auth: { id: 51, privateKey: readFileSync("private-key.pem") },
  baseUrl: "https://ghe.io/api/v3",
});

octokit.request("GET /app").then(console.log, console.error);
```

But unfortunately it worked without a hitch. Here is what I did

- I created a new GitHub app without any extra permission and set the webhook URL to `https://example.com/`.
- I've downloaded the `*.pem` file, renamed it to `private-key.pem` and saved it in the same directory as my test script above
- I set the `auth.id` to the app ID and `baseUrl` to the GHE instances' API base URL.

I'm a bit out of ideas. Maybe something is wrong with the `privateKey` formatting when you define it inline in the code? Coudl you try to use `readFileSync` instead?
